File: scala/triangle/src/main/scala/Triangle.scala
case class Triangle(s1: Double, s2: Double, s3: Double) {

  val l = List(s1, s2, s3).sorted
  val s = l.toSet

  val isValid = !s.exists(_ <= 0) && l.head + l(1) >= l(2)
  def scalene = isValid && s.size == 3
  def isosceles = isValid && s.size <= 2
  def equilateral = isValid && s.size == 1
}

File: haskell/triangle/src/Triangle.hs
module Triangle (TriangleType(..), triangleType) where
import Data.List

data TriangleType = Equilateral
                  | Isosceles
                  | Scalene
                  | Illegal
                  deriving (Eq, Show)

triangleType :: Float -> Float -> Float -> TriangleType
triangleType a b c
    | x + y < z || x == 0 = Illegal
    | x == y && y == z = Equilateral
    | x == y || y == z = Isosceles
    | otherwise = Scalene
    where
      l = sort [a, b, c]
      x = l !! 0
      y = l !! 1
      z = l !! 2
